src/main/java/upeldev/com/github/upel3/services/CourseService.java

```diff
@@ -68,5 +70,23 @@ public void addLecturer(Course course, User lecturer){
         userRepository.save(lecturer);
     }
 
+    public Course findCourseByAccessCode(String accessCodeString){
+        Optional<AccessCode> accessCodeOptional = accessCodeRepository.findById(accessCodeString);
+        if (accessCodeOptional.isEmpty())
+            return null;
+        AccessCode accessCode = accessCodeOptional.get();
+        return courseRepository.findByAccessCode(accessCode);
+    }
+
+    public void addStudent(User student, String courseCode){
+        Course course = findCourseByAccessCode(courseCode);
+        if (course == null){
+            throw new IllegalArgumentException("No such course");
+        }
+        if (course.getEnrolledStudents().contains(student)){
+            throw new IllegalArgumentException("Student it already enrolled to this course");
+        }
+        addStudent(course, student);
```

Comment:
Rename one of the methods or make addStudent that doesn't throw exceptions private.